k9mail/src/main/java/com/fsck/k9/mailstore/ProvidedTempFileBody.java

```diff
@@ -82,11 +88,11 @@ public long getSize() {
         throw new IllegalStateException("Data must be fully written before it can be read!");
     }
 
-    public File getFile() throws IOException {
+    public Uri getProviderUri(String mimeType) throws IOException {
```

Comment:
The Body only needs to know that it needs to write to a file. Similarly, the file provider only cares about files and content URIs, not message bodies. The thing that knows that these two things can be combined to achieve the desired result should live a layer above the other two.
We violate this separation in a lot of different places. So it's not that easy to get it right here. But it leads to really ugly class names :)